src/exchange.ts

```diff
@@ -185,26 +185,35 @@ export const cacheExchange = (opts?: CacheExchangeOpts): Exchange => ({
   };
 
   // Take any OperationResult and update the cache with it
-  const updateCacheWithResult = ({ data, operation }: OperationResult) => {
-    let dependencies;
-    if (data !== null && data !== undefined) {
-      dependencies = write(store, operation, data).dependencies;
-    }
-
+  const updateCacheWithResult = (result: OperationResult) => {
+    const { data, operation } = result;
+    let writeDependencies, queryDependencies;
     // Clear old optimistic values from the store
     const { key } = operation;
     if (optimisticKeys.has(key)) {
       optimisticKeys.delete(key);
       store.clearOptimistic(key);
     }
 
-    if (dependencies !== undefined) {
+    if (data !== null && data !== undefined) {
+      writeDependencies = write(store, operation, data).dependencies;
+      if (isQueryOperation(operation)) {
+        const queryResult = query(store, operation);
+        result.data = queryResult.data;
+        queryDependencies = queryResult.dependencies;
+      } else {
+        const res = readOperation(store, operation, data);
+        result.data = res.data;
+      }
+    }
+
+    if (writeDependencies !== undefined) {
```

Comment:
Lets maybe separate this. We're only checking for writeDependencies so below we should maybe check separately for readDependencies?